content/blog/2022-04-19-tracing/index.md

```diff
@@ -0,0 +1,142 @@
+1. For each program, the vast majority of execution time is spent on loops. So that optimizing loops is often more fruitful.
+2. Loop traces are often type-stable, even more so than methods. A variable in loop start with integer would almost always remain as integer for all iterations.
+
+We would mainly take a look at three aspects of TraceMonkey:
+1. Design Overview && Procedures
+2. Trace/Tracetree Formation
+3. TraceMonkey VM Implementation && Optimizations
+
+# Design Overview
+Below is a state diagram from the original paper showing the overall structure of TraceMonkey. To make the explanation clearer, I've also attached the sample program: Sieve of Eratosthenes, which is also from the original paper.
+
+<p align="center">
+<img src="state_diagram.png" alt="state_diagram" title="state_diagram" style="zoom:25%;" />
+</p>
+
+
+```
+for (var i = 2; i < 100; ++i) {
+    if (!primes[i])
+        continue;
+    for (var k = i + i; i < 100; k += i)
+        primes[k] = false
+}
+```
+
+Each program in tracemonkey starts with running in interpreting mode. The trace points in TraceMonkey are the loop edges. Every time when the interpreter crosses the loop edge, the trace monitor gets invoked, which would then decide in between continuing with interpreting mode, recording a new trace, or executing a compiled trace. At the beginning, none of the traces are hot, so the monitor will always pick interpreting mode. In the context of the sample program, this is when `i == 2` and we are first entering the outer loop. After we execute the inner program for the second time, the trace monitor would identify that loop as hot and would start recording. During the recording, the program will be executed in interpreted mode and the trace is recorded with a low-level compiler intermediate representation called LIR. LIR encodes all operations with type information, as well as guards and stack operations when crossing a function boundary. The LIR snippet for encoding the inner loop is attached below:
+
+```
+    v0 := ld state[748] // load primes from the trace activation record
+    st sp[0], v0 // store primes to interpreter stack
+    v1 := ld state[764] // load k from the trace activation record
+    v2 := i2f(v1) // convert k from int to double
+    st sp[8], v1 // store k to interpreter stack
+    st sp[16], 0 // store false to interpreter stack
+    v3 := ld v0[4] // load class word for primes
+    v4 := and v3, -4 // mask out object class tag for primes
+    v5 := eq v4, Array // test whether primes is an array
+    xf v5 // side exit if v5 is false
+    v6 := js_Array_set(v0, v2, false) // call function to set array element
+    v7 := eq v6, 0 // test return value from call
+    xt v7 // side exit if js_Array_set returns false
```

Comment:
FWIW, no need to *both* indent code *and* surround it with backticks. Both do the same thing in Markdown (mark the text as a code block). Doing both will make it into indented code in the final output.